### PR TITLE
[WebPubSub]Implement WebSocket Client

### PR DESCRIPTION
**Pull request by zackliu**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:48]:
```suggestion
        public virtual void ClientStateChanged(string newState, string currentState)
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:84]: Use a consistent pattern, FailedToDo or FailedDoing?

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:111]: Failed...

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:45]: is there any scenario when the message is not endOfMessage?

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:47]: do you want a fast version trying .Wait()

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:67]: interesting.. do you want it to be thread safe

Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:67]: interesting.. do you want it to be thread safe

Comment (vicancy) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:67]: is using a local var better?

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClientFactory.cs
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClientFactory.cs:12]: why need factory?

Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClientFactory.cs:12]: For test.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:67]: It's thread safe. buffer uses `ThreadStatic`

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:45]: No currently.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:47]: Make it an optimization in future

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs:67]: This doesn't have to be thread safe as no concurrent invocation

**Review by jsquire — CHANGES_REQUESTED** [review]
As part of these changes, let's please adjust the target framework to `netstandard2.0` to conform to [guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html#azure-sdk-library-design).   In addition to this blocking any release, it's also causing some test pipelines to break.

We'll want to remove the override to `<RequiredTargetFrameworks>` that was made [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj#L9).    We'll also need to fix the test project to remove the condition to avoid `net461` that was added [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj#L15) and ensure that tests are passing across the supported Azure SDK targets.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj
Comment (jsquire) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj:9]: As a general rule, we'll want this package to have the same targets as the other Azure SDK packages - meaning that we should have just the `$(RequiredTargetFrameworks)` here.

If you have a strong reason to target `netstandard2.1` explicitly, we'll need to discuss with the architects and approval.  Can you help me understand the rationale for this, please?

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj
Comment (jsquire) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj:14]: This condition needs to be removed.   Tests must run and pass on all of the frameworks supported by Azure SDK packages.

Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj:14]: Will be removed.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj:9]: The consider here is if I target `netstandard2.1` explicitly, then I can use some `netstandard2.1` only interface/features. E.g. `Memory<byte>` overload in WebSocket client, which can improve the performance and reduce GCs. As it's a data plane SDK, we need consider the performance issue.

Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj]: I'm temporarily reverting the code for troubleshooting the ci failure issue.

Comment (jsquire) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj]: As mentioned, these are standardized across the Azure SDK packages and not something that we generally allow individual packages to deviate from.   We are having discussions about potentially adding `net6.0` in the near future, which would unlock those same multi-targeting scenarios.

If including `netstandard2.1` is something that you feel strongly about, it will require architect approval for an exception.  Please open a thread with @KrzysztofCwalina, @tg-msft, and myself to discuss.   Until that exception is approved, please keep this as `$(RequiredTargetFrameworks)` and remove any overrides.

Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Azure.Messaging.WebPubSub.Client.csproj]: @jsquire I don't think there's a very strong justification now. I'll remove the overrides and looking forward adding net6.0

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Internals/MemoryBufferWriter.cs
Comment (weshaggard) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/Internals/MemoryBufferWriter.cs:200]: Your library is no longer target .NET STANDARD 2.1 or higher. RequiredTargetFrameworks is set to netstandard2.0 only now with your change to the project.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj
Comment (weshaggard) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj:15]: @zackliu this change could also potentially account for the test failures only on net461. If there is something in that this library depends on that doesn't work on net461.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj
Comment (jsquire) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/tests/Azure.Messaging.WebPubSub.Client.Tests.csproj:6]: The standard for Azure SDK libraries is NUnit; I overlooked the xunit reference.  We'll need to move tests over.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:38]:
```suggestion
        [Event(1, Level = EventLevel.Verbose, Message = "Client is starting.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:47]:
```suggestion
        [Event(2, Level = EventLevel.Verbose, Message = "The client state changes from the '{0}' state to the '{1}' state.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:56]:
```suggestion
        [Event(3, Level = EventLevel.Verbose, Message = "A new WebSocket connection starts to connect with subprotocol {0}.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:65]:
```suggestion
        [Event(4, Level = EventLevel.Verbose, Message = "WebSocket connection is closed.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:74]:
```suggestion
        [Event(5, Level = EventLevel.Warning, Message = "An exception is occurred while processing message from the service. Error Message: {0}")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:38]: Shall we make connection start/stop to __Information__ level cause it's kind of stability.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:83]:
```suggestion
        [Event(6, Level = EventLevel.Informational, Message = "An exception is occurred while receiving bytes. Error Message: {0}")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:92]:
```suggestion
        [Event(7, Level = EventLevel.Warning, Message = "The client failed to change from the '{0}' state to the '{1}' state because it was actually in the '{2}' state.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:101]:
```suggestion
        [Event(8, Level = EventLevel.Warning, Message = "Stop trying to recover the connection with connection ID: {0} cause of {1}.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:110]:
```suggestion
        [Event(9, Level = EventLevel.Informational, Message = "An attempt to recover connection with connection ID: {0} is failed, will retry later. Error Message: {1}")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:119]:
```suggestion
        [Event(10, Level = EventLevel.Warning, Message = "An exception occurred while invoking event handler {0}. Error Message: {1}")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:128]:
```suggestion
        [Event(11, Level = EventLevel.Informational, Message = "Connection with connection ID: {0} is connected.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:137]:
```suggestion
        [Event(12, Level = EventLevel.Informational, Message = "Connection with connection ID: {0} is disconnected.")]
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:146]:
```suggestion
        [Event(13, Level = EventLevel.Warning, Message = "An attempt to reconnect connection who is the successor of connection ID: {0} is failed, will retry later. Error Message: {1}")]
```

**Review by JialinXin — APPROVED** [review]

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (jsquire) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:74]: The `is` should not be added here, as it is not proper grammar.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (jsquire) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:56]: This should probably be `started` or `is starting`

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (jsquire) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:47]: This should be `changed` or `is changing`

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (jsquire) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:110]: The original version is proper grammar; the suggestion is not.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs
Comment (jsquire) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/EventSource/WebPubSubClientEventSource.cs:146]: The original version is proper grammar; the suggestion is not.

**Review by jsquire — APPROVED** [review]
Structure, references, and other guidelines-related changes look good.

**Pull request merged**
